File: inst/script/makeExampleData.py
#!/usr/bin/env python3
"""Generate the example peak files shipped in inst/extdata.

Three replicates over a stretch of chr1 and chr2. Most positions are
shared, a portion is deliberately weak in one replicate so the rescue
step has something to do, and the third replicate carries extra noise
peaks so the discard step does too. A BED3 copy of the first replicate is
written as well, to exercise the presence-only path.

Run with: python3 makeExampleData.py ../extdata
"""

import random
import sys
from pathlib import Path

random.seed(20240517)

N_SHARED = 220
N_WEAK = 60
N_NOISE = 90
CHROMS = {"chr1": 12_000_000, "chr2": 9_000_000}


def make_positions(n, used, chrom_lengths):
    """Draw non-overlapping peak starts, keeping a gap between them."""
    out = []
    while len(out) < n:
        chrom = random.choice(list(chrom_lengths))
        start = random.randint(1, chrom_lengths[chrom] - 2000)
        block = start // 5000
        if (chrom, block) in used:
            continue
        used.add((chrom, block))
        out.append((chrom, start))
    return out


def jitter(start, width, amount=90):
    """Shift a peak slightly, the way independent calls never agree exactly."""
    shift = random.randint(-amount, amount)
    return max(1, start + shift), width + random.randint(-40, 40)


def write_narrowpeak(path, records):
    with open(path, "w") as handle:
        for i, (chrom, start, width, neglog10p) in enumerate(records):
            end = start + width
            summit = width // 2 + random.randint(-30, 30)
            score = min(1000, int(neglog10p * 12))
            handle.write(
                f"{chrom}\t{start}\t{end}\tpeak_{i + 1}\t{score}\t.\t"
                f"{neglog10p / 3:.5f}\t{neglog10p:.5f}\t"
                f"{max(0.0, neglog10p - 1):.5f}\t{summit}\n"
            )


def write_bed3(path, records):
    with open(path, "w") as handle:
        for chrom, start, width, _ in records:
            handle.write(f"{chrom}\t{start}\t{start + width}\n")


def build():
    used = set()
    shared = make_positions(N_SHARED, used, CHROMS)
    weak = make_positions(N_WEAK, used, CHROMS)

    replicates = []
    # the third replicate is the shallow one: weaker everywhere and noisier
    for depth in (1.0, 0.95, 0.55):
        records = []

        for chrom, start, in shared:
            width = random.randint(280, 900)
            s, w = jitter(start, width)
            strength = random.gauss(11, 3) * depth
            records.append((chrom, s, max(w, 150), max(strength, 4.2)))

        # positions that only clear the weak threshold anywhere
        for chrom, start in weak:
            width = random.randint(250, 600)
            s, w = jitter(start, width)
            strength = random.gauss(5.2, 0.9) * depth
            records.append((chrom, s, max(w, 150), max(strength, 4.05)))

        # replicate-specific noise, no support expected
        noise_used = set()
        for chrom, start in make_positions(N_NOISE, noise_used, CHROMS):
            width = random.randint(200, 450)
            strength = max(random.gauss(4.9, 0.6), 4.02)
            records.append((chrom, start, width, strength))

        records.sort(key=lambda r: (r[0], r[1]))
        replicates.append(records)

    return replicates


def main():
    target = Path(sys.argv[1] if len(sys.argv) > 1 else "../extdata")
    target.mkdir(parents=True, exist_ok=True)

    replicates = build()
    for i, records in enumerate(replicates, start=1):
        write_narrowpeak(target / f"rep{i}.narrowPeak", records)

    write_bed3(target / "rep1_minimal.bed", replicates[0])
    print(f"wrote {len(replicates)} narrowPeak files to {target}")


if __name__ == "__main__":
    main()

File: inst/script/makeExampleData.R
#!/usr/bin/env Rscript

## Generate the example peak files shipped in inst/extdata.
##
## Three replicates over two invented chromosomes. Most positions are
## shared, a portion is deliberately left too weak for any single
## replicate to confirm so that the rescue step has something to do, and
## each replicate carries its own noise peaks so that the discard step
## does too. A BED3 copy of the first replicate is written as well, to
## exercise the path taken when the input carries no statistic at all.
##
## Run from this directory with:
##   Rscript makeExampleData.R ../extdata

set.seed(20240517)

nShared <- 220L
nWeak <- 60L
nNoise <- 90L
nChained <- 12L
chainLength <- 5L
blockSize <- 5000L
chromosomeLengths <- c(chr1 = 12000000L, chr2 = 9000000L)

## Peaks are placed one per 5 kb block, which keeps the planted positions
## clearly separated without needing a rejection loop. Blocks are drawn
## from a single shuffled pool so that no position is ever reused, and
## each chromosome contributes in proportion to its length.
blockPool <- do.call(rbind, lapply(names(chromosomeLengths), function(chr) {
    data.frame(chromosome = chr,
               block = seq_len(chromosomeLengths[[chr]] %/% blockSize) - 1L,
               stringsAsFactors = FALSE)
}))
blockPool <- blockPool[sample.int(nrow(blockPool)), ]
blocksTaken <- 0L

takePositions <- function(n, maxOffset = blockSize - 1500L) {
    selected <- blockPool[blocksTaken + seq_len(n), ]
    blocksTaken <<- blocksTaken + n

    ## a random offset inside the block, leaving room for the widest peak
    ## that will be placed there
    data.frame(chromosome = selected$chromosome,
               start = selected$block * blockSize +
                   sample.int(maxOffset, n, replace = TRUE),
               stringsAsFactors = FALSE)
}

## Independent peak calls never agree on a boundary to the base pair, so
## the shared positions are shifted a little in every replicate.
jitterPeaks <- function(start, width, amount = 90L) {
    data.frame(
        start = pmax(1L, start + sample(-amount:amount, length(start),
                                        replace = TRUE)),
        width = pmax(150L, width + sample(-40:40, length(width),
                                          replace = TRUE)))
}

writeNarrowPeak <- function(peaks, path) {
    ## the ENCODE ten-column layout: the eighth column carries -log10 p,
    ## which is what readPeakSets() picks up
    summit <- round(peaks$width / 2) +
        sample(-30:30, nrow(peaks), replace = TRUE)

    table <- data.frame(
        chromosome = peaks$chromosome,
        start = peaks$start,
        end = peaks$start + peaks$width,
        name = paste0("peak_", seq_len(nrow(peaks))),
        score = pmin(1000L, as.integer(peaks$negLog10P * 12)),
        strand = ".",
        signalValue = sprintf("%.5f", peaks$negLog10P / 3),
        pValue = sprintf("%.5f", peaks$negLog10P),
        qValue = sprintf("%.5f", pmax(0, peaks$negLog10P - 1)),
        peak = summit,
        stringsAsFactors = FALSE)

    utils::write.table(table, file = path, sep = "\t", quote = FALSE,
                       row.names = FALSE, col.names = FALSE)
}

writeBed3 <- function(peaks, path) {
    table <- data.frame(chromosome = peaks$chromosome,
                        start = peaks$start,
                        end = peaks$start + peaks$width,
                        stringsAsFactors = FALSE)

    utils::write.table(table, file = path, sep = "\t", quote = FALSE,
                       row.names = FALSE, col.names = FALSE)
}

## positions every replicate draws from
sharedPositions <- takePositions(nShared)
weakPositions <- takePositions(nWeak)

## A handful of loci where each replicate lays down a short run of
## partially overlapping peaks: the first and last of a run do not touch,
## so transitive merging chains them into one long region while the
## seeded rule does not. Without these the two merge methods would give
## identical output and there would be nothing to demonstrate.
chainPositions <- takePositions(nChained, maxOffset = 1000L)

## the third replicate stands in for a shallower library: the same peaks
## are there, but weaker, so it leans on the other two to confirm them
depthFactors <- c(1.00, 0.95, 0.55)
replicates <- lapply(depthFactors, function(depth) {
    sharedWidth <- sample(280:900, nShared, replace = TRUE)
    sharedShift <- jitterPeaks(sharedPositions$start, sharedWidth)
    shared <- data.frame(
        chromosome = sharedPositions$chromosome,
        start = sharedShift$start,
        width = sharedShift$width,
        negLog10P = pmax(4.2, stats::rnorm(nShared, 11, 3) * depth),
        stringsAsFactors = FALSE)

    ## these clear the weak threshold and nothing more, in any replicate
    weakWidth <- sample(250:600, nWeak, replace = TRUE)
    weakShift <- jitterPeaks(weakPositions$start, weakWidth)
    weak <- data.frame(
        chromosome = weakPositions$chromosome,
        start = weakShift$start,
        width = weakShift$width,
        negLog10P = pmax(4.05, stats::rnorm(nWeak, 5.2, 0.9) * depth),
        stringsAsFactors = FALSE)

    ## replicate-specific noise, which nothing else should support
    noisePositions <- takePositions(nNoise)
    noise <- data.frame(
        chromosome = noisePositions$chromosome,
        start = noisePositions$start,
        width = sample(200:450, nNoise, replace = TRUE),
        negLog10P = pmax(4.02, stats::rnorm(nNoise, 4.9, 0.6)),
        stringsAsFactors = FALSE)

    ## each step advances by less than one peak width, so neighbours
    ## overlap but the ends of a run do not
    chainWidth <- 600L
    chainStep <- 400L
    chained <- do.call(rbind, lapply(seq_len(chainLength), function(step) {
        data.frame(
            chromosome = chainPositions$chromosome,
            start = chainPositions$start + (step - 1L) * chainStep,
            width = chainWidth,
            negLog10P = pmax(4.2, stats::rnorm(nChained, 12, 2) * depth),
            stringsAsFactors = FALSE)
    }))

    peaks <- rbind(shared, weak, noise, chained)
    peaks[order(peaks$chromosome, peaks$start), ]
})

target <- commandArgs(trailingOnly = TRUE)
target <- if (length(target) > 0) target[1] else "../extdata"
dir.create(target, showWarnings = FALSE, recursive = TRUE)

for (i in seq_along(replicates)) {
    writeNarrowPeak(replicates[[i]],
                    file.path(target, paste0("rep", i, ".narrowPeak")))
}

## the same coordinates with every statistic stripped away
writeBed3(replicates[[1]], file.path(target, "rep1_minimal.bed"))

message("wrote ", length(replicates), " narrowPeak files to ", target)
